databus/src/main/java/com/bazaarvoice/emodb/databus/db/generic/CachingSubscriptionDAO.java

```diff
@@ -25,65 +40,288 @@
  * Wraps a {@link SubscriptionDAO} with a cache that makes it fast and efficient to lookup subscription metadata.  The
  * downside is that servers must globally coordinate changes to subscriptions because the consequences of using
  * out-of-date cached subscription metadata are pretty severe.
+ *
+ * There have been two implementations for how cache invalidation is managed using the cache registry:
+ *
+ * <ul>
+ *     <li>
+ *         The current implementation registers a cache named "subscriptionsByName" which maps subscription names
+ *         to OwnedSubscriptions.  When a subscription is changed only the corresponding subscription is invalidated.
+ *     </li>
+ *     <li>
+ *         The legacy implementation registered a cache named "subscriptions" which mapped a single key, also named
+ *         "subscriptions", to a complete map of subscription names to OwnedSubscriptions.  When a subscription was
+ *         changed the single value in the cache was invalidated, causing all subscriptions to be reloaded and
+ *         cached as a map.
+ *     </li>
+ * </ul>
+ *
+ * Over time, as the number of total subscriptions, single-subscription lookups and cache invalidations scaled up the
+ * legacy caching became a bottleneck and was replaced with the current one.  However, since the two cache invalidation
+ * systems are not compatible it is not possible to just upgrade from one to another on an in-flight system without
+ * either incurring downtime or potentially losing invalidation messages.  Therefore to support in-flight upgrades
+ * there is a {@link CachingMode} parameter which controls how cache invalidations are managed.
+ *
+ * <ol>
+ *     <li>
+ *         In legacy mode the DAO uses legacy cache invalidation.  It listens for invalidation messages and posts them
+ *         to the legacy cache.  It also listens for current cache invalidation, although none will be posted until
+ *         the next phase.
+ *     </li>
+ *     <li>
+ *         In bridge mode the DAO still listens for legacy cache invalidations but only posts current cache invalidations.
+ *     </li>
+ *     <li>
+ *         In normal mode the DAO exclusively uses current cache invalidation.
+ *     </li>
+ * </ol>
+ *
+ * A safe upgrade requires upgrading all servers from legacy to bridge to normal mode.  Only once all active servers
+ * are in one mode is it safe to move to the next.
  */
 public class CachingSubscriptionDAO implements SubscriptionDAO {
 
+    public enum CachingMode {
+        legacy,
+        bridge,
+        normal
+    };
+
```

Comment:
Bridge mode seem to be adding some complexity. As we discussed, I liked just having two modes: legacy and normal, with legacy mode listening from both _legacyCache and _subscriptionCache and posting invalidations on both of them for the time till we’re ready to switch from legacy to Normal.